src/volatile_memory.rs

```diff
@@ -146,6 +178,61 @@ pub trait VolatileMemory {
         }
     }
 
+    /// Gets a reference to T at `offset`.  The resulting pointer must be
+    /// aligned, or the function fails.
+    ///
+    /// To use this safely, the caller must guarantee that there are no other
+    /// users of the given chunk of memory for the lifetime of the result.
+    unsafe fn aligned_as_ref<T: ByteValued>(&self, offset: usize) -> Result<&T> {
+        let slice = self.get_slice(offset, size_of::<T>())?;
+
+        if ((slice.addr as usize) & (align_of::<T>() - 1)) != 0 {
+            Err(Error::Misaligned {
+                addr: slice.addr as usize,
+                alignment: align_of::<T>(),
+            })
+        } else {
+            Ok(&*(slice.addr as *const T))
+        }
+    }
+
+    /// Gets a reference to T at `offset`.  The resulting pointer must be
+    /// aligned, or the function fails.
+    ///
+    /// To use this safely, the caller must guarantee that there are no other
+    /// users of the given chunk of memory for the lifetime of the result.
+    unsafe fn aligned_as_mut<T: ByteValued>(&self, offset: usize) -> Result<&mut T> {
+        let slice = self.get_slice(offset, size_of::<T>())?;
+
+        if ((slice.addr as usize) & (align_of::<T>() - 1)) != 0 {
+            Err(Error::Misaligned {
+                addr: slice.addr as usize,
+                alignment: align_of::<T>(),
+            })
+        } else {
+            Ok(&mut *(slice.addr as *mut T))
+        }
+    }
+
+    /// Gets a reference to T at `offset`.  The resulting pointer must be
+    /// aligned, or the function fails.
+    fn get_atomic_ref<T: AtomicValued>(&self, offset: usize) -> Result<&T> {
```

Comment:
I was wondering, isn't atomicity orthogonal to pointer aliasing? If so, shouldn't this function be marked as `unsafe`, with the same contract as the previous two (*to use this safely, the caller must guarantee that there are no other users of the given chunk of memory for the lifetime of the result*)?  Moreover, in that case, does this become equivalent to `aligned_as_ref` if `AtomicValued: ByteValued`?

Reply:
Actually the point of atomic types is _exactly_ that they can be aliased safely.

Atomic types, if you ignore all the crazy parts with `std::sync::atomic::Ordering`, behave very similarly to an int type wrapped with a mutex, e.g. `AtomiicU8` -> `Mutex<u8>`. In particular, atomic types provide interior mutability just like `Mutex`, i.e. they can be accessed concurrently even in the presence of aliases. This can be done because concurrent accesses to atomic types do not create a [data race](https://doc.rust-lang.org/nomicon/races.html) (yes, I'm showing off by linking the nomicon but you don't have to read it if you trust my comparison with `Mutex`:)), 

"Not creating a data race" of course says nothing about correctness; it's only about safety. However that should be enough to answer your question..

Reply:
Thanks for the explanation! I don't think linking to anything is showing off; it's great to be pointed towards more information. I was initially worried about interpreting the same piece of memory as both atomic using this method, and also as some non-atomic value in some other place in the code, but I now realized this would be a breakage of the unsafe contract for the non-atomic access so yeah looks fine. 

Last question: so you're basically saying that data races are the only reason why pointer aliasing can lead to undefined behavior? (I promise I'll have a much better grasp of memory models some day)

Reply:
> also as some non-atomic value in some other place in the code, but I now realized this would be a breakage of the unsafe contract for the non-atomic access so yeah looks fine.

Actually even if you did, safety would be guaranteed by the contract of the `ByteValued` trait together with usage of `read_volatile`/`write_volatile`, but I agree that "don't do it" is the best thing to do with respect to mixing atomic and non-atomic accesses.

> so you're basically saying that data races are the only reason why pointer aliasing can lead to undefined behavior

Yes, pretty much! Again, this is only talking about undefined behavior, correctness is of course a whole other level of complexity: data races is why aliasing `Rc<T>` leads to undefined behavior, but there's more to `Arc<T>` than just making the reference count atomic.